### DISCUSSION TITLE
Question: Pre-install extensions with npm (automation-script)

### DISCUSSION THREAD

**Discussion by r10r** [neoclide/coc.nvim#3733] [Q&A] [answered]
Hi,

I'm a new user of coc.nvim and want to create a docker image with extensions preinstalled.
I followed https://github.com/neoclide/coc.nvim/wiki/Install-coc.nvim#automation-script 
to install extensions with `npm` but the installed extensions do not show up in 
the output of `:CocList extensions`.

Is there another way to pre-install extensions (system-wide) for air-gapped environments ?

Thanks
Ruben

**Comment by chemzqm** [comment, marked as answer]
Check package.json and node_modules in the folder `:echo coc#util#get_data_home().'/extensions'`

**Comment by r10r** [comment]
@chemzqm  Thanks for the response.
 
The problem for me is that *$HOME* is located on a volume, 
so modules can't be pre-installed to *$HOME/.config/coc/extensions* in the container image.

But I figured out that `npm link` can be used instead of `npm install`

1. Install extension node modules globally in container image:

```Dockerfile
# install coc extension (node module) globally
RUN npm install -g coc-snippets --global-style --ignore-scripts --no-bin-links --no-package-lock --only=prod
```

2. Link extensions in the containers **entrypoint.sh**

```sh
# see https://github.com/neoclide/coc.nvim/wiki/Install-coc.nvim#automation-script
mkdir -p ~/.config/coc/extensions
cd ~/.config/coc/extensions
if [ ! -f package.json ]
then
  echo '{"dependencies":{}}'> package.json
fi
# link globally installed extension
npm link coc-snippets --save --no-package-lock --only=prod
```

* Pre-installed modules are installed to */usr/local/lib/node_modules*
* Extensions installed by the user will be installed to *~/.config/coc/extensions/node_modules*
